File: packages/swift/Sources/IdentifoldPostgres/PostgresStorageAdapter.swift
import Identifold
import PostgresNIO

@available(macOS 13.0, iOS 16.0, tvOS 16.0, watchOS 9.0, *)
public struct PostgresStorageAdapter: StorageAdapter, Sendable {
  private let client: PostgresClient

  public init(client: PostgresClient) {
    self.client = client
  }

  public func reserve(_ request: ReferenceReservation) async throws -> Bool {
    try await databaseOperation {
      let rows = try await client.query("""
        SELECT identifold_reserve_reference(
          \(request.machineID)::text::uuid,
          \(request.namespace)::text,
          \(request.reference)::text
        )
        """)
      var iterator = rows.decode(Bool.self).makeAsyncIterator()
      guard let reserved = try await iterator.next(), try await iterator.next() == nil else {
        throw IdentifoldError("allocation_conflict")
      }
      return reserved
    }
  }

  public func resolve(reference: String, namespace: String) async throws -> ReferenceMapping? {
    try await databaseOperation {
      let rows = try await client.query("""
        SELECT resolved_machine_id::text, resolved_namespace
        FROM identifold_resolve_reference(\(reference)::text, \(namespace)::text)
        """)
      var iterator = rows.decode((String, String).self).makeAsyncIterator()
      guard let value = try await iterator.next() else { return nil }
      guard try await iterator.next() == nil else {
        throw IdentifoldError("allocation_conflict")
      }
      return ReferenceMapping(machineID: value.0, namespace: value.1)
    }
  }

  public func allocate(_ request: SequenceAllocationRequest) async throws -> UInt64 {
    try await databaseOperation {
      let rows: PostgresRowSequence
      if let scope = request.scope {
        rows = try await client.query("""
          SELECT identifold_allocate_sequence(
            \(request.machineID)::text::uuid,
            \(request.namespace)::text,
            \(request.referencePrefix)::text,
            \(scope)::text,
            \(Int16(request.width))::smallint
          )
          """)
      } else {
        rows = try await client.query("""
          SELECT identifold_allocate_sequence(
            \(request.machineID)::text::uuid,
            \(request.namespace)::text,
            \(request.referencePrefix)::text,
            NULL::text,
            \(Int16(request.width))::smallint
          )
          """)
      }
      var iterator = rows.decode(Int64.self).makeAsyncIterator()
      guard let value = try await iterator.next(), value >= 0,
        try await iterator.next() == nil
      else {
        throw IdentifoldError("allocation_conflict")
      }
      return UInt64(value)
    }
  }

  private func databaseOperation<T: Sendable>(
    _ operation: () async throws -> T
  ) async throws -> T {
    do {
      return try await operation()
    } catch let error as IdentifoldError {
      throw error
    } catch let error as PSQLError {
      switch error.serverInfo?[.sqlState] {
      case "22003": throw IdentifoldError("sequence_overflow")
      case "22023": throw IdentifoldError("invalid_allocation_policy")
      default: throw IdentifoldError("allocation_conflict")
      }
    } catch {
      throw IdentifoldError("allocation_conflict")
    }
  }
}

File: packages/csharp/src/Identifold/Storage/Postgres/PostgresStorageAdapter.cs
using Npgsql;
using NpgsqlTypes;

namespace Greyfoundry.Identifold.Storage.Postgres;

public sealed class PostgresStorageAdapter(NpgsqlDataSource dataSource) : IStorageAdapter
{
    public async ValueTask<bool> ReserveAsync(
        ReferenceReservation request,
        CancellationToken cancellationToken = default)
    {
        try
        {
            await using var command = dataSource.CreateCommand(
                "SELECT identifold_reserve_reference($1::text::uuid, $2::text, $3::text)");
            command.Parameters.AddWithValue(request.MachineId);
            command.Parameters.AddWithValue(request.Namespace);
            command.Parameters.AddWithValue(request.Reference);
            var value = await command.ExecuteScalarAsync(cancellationToken);
            return value is bool reserved
                ? reserved
                : throw new IdentifoldException("allocation_conflict");
        }
        catch (IdentifoldException) { throw; }
        catch (PostgresException exception) { throw Map(exception); }
        catch (NpgsqlException) { throw new IdentifoldException("allocation_conflict"); }
    }

    public async ValueTask<ReferenceMapping?> ResolveAsync(
        string reference,
        string namespaceName,
        CancellationToken cancellationToken = default)
    {
        try
        {
            await using var command = dataSource.CreateCommand(
                "SELECT resolved_machine_id::text, resolved_namespace " +
                "FROM identifold_resolve_reference($1::text, $2::text)");
            command.Parameters.AddWithValue(reference);
            command.Parameters.AddWithValue(namespaceName);
            await using var reader = await command.ExecuteReaderAsync(cancellationToken);
            if (!await reader.ReadAsync(cancellationToken)) return null;
            return new ReferenceMapping(reader.GetString(0), reader.GetString(1));
        }
        catch (PostgresException exception) { throw Map(exception); }
        catch (NpgsqlException) { throw new IdentifoldException("allocation_conflict"); }
    }

    public async ValueTask<ulong> AllocateAsync(
        SequenceAllocationRequest request,
        CancellationToken cancellationToken = default)
    {
        try
        {
            await using var command = dataSource.CreateCommand(
                "SELECT identifold_allocate_sequence(" +
                "$1::text::uuid, $2::text, $3::text, $4::text, $5::smallint)");
            command.Parameters.AddWithValue(request.MachineId);
            command.Parameters.AddWithValue(request.Namespace);
            command.Parameters.AddWithValue(request.ReferencePrefix);
            command.Parameters.Add(new NpgsqlParameter
            {
                NpgsqlDbType = NpgsqlDbType.Text,
                Value = request.Scope is null ? DBNull.Value : request.Scope,
            });
            command.Parameters.AddWithValue((short)request.Width);
            var value = await command.ExecuteScalarAsync(cancellationToken);
            return value is long sequence && sequence >= 0
                ? (ulong)sequence
                : throw new IdentifoldException("allocation_conflict");
        }
        catch (IdentifoldException) { throw; }
        catch (PostgresException exception) { throw Map(exception); }
        catch (NpgsqlException) { throw new IdentifoldException("allocation_conflict"); }
    }

    private static IdentifoldException Map(PostgresException exception) => exception.SqlState switch
    {
        "22003" => new("sequence_overflow"),
        "22023" => new("invalid_allocation_policy"),
        _ => new("allocation_conflict"),
    };
}
